### PR TITLE
fix(mp): reload cached KV for preempted requests

### PR DESCRIPTION
**Pull request by Carlos779988**
**What this PR does / why we need it**:

`LMCacheMPConnector.get_num_new_matched_tokens()` currently returns `(0, False)` for every preempted request. A resumed request therefore recomputes its prompt and generated-token prefix even when those KV chunks are available in LMCache.

This PR removes that early return and lets preempted requests use the existing MP lookup/load path. The connector already rebuilds the request tracker after preemption from `request.all_token_ids`, so the lookup covers both prompt and generated tokens.

Scope: this PR only enables reload of KV chunks that are already durable in LMCache. It does not change STORE completion ordering or GPU block lifetime, and it should not be treated as making arbitrary or cooperative preemption storage-safe. STORE lifetime safety and broader worker/metadata lifecycle coverage will be addressed in a separate follow-up PR after aligning on the operation-level contract.

For preempted full-sequence hits, this PR leaves the final token for vLLM to recompute logits. #3322 proposes the same last-token handling for the general full-prompt-hit path.

Refs #2007 and #2929.

Added scheduler-side tests for:

- loading a cached prompt/output prefix after preemption;
- polling an asynchronous lookup without resetting the fresh tracker;
- recomputing the final token on a preempted full-sequence hit.

**Special notes for your reviewers**:

Validated on an NVIDIA H20 against LMCache `48dd4df6` with vLLM `0.27.2rc1.dev150+g311b3513a`:

- Qwen3-0.6B: the unmodified base reported no LMCache recovery after preemption; this branch recovered 2048-2304 tokens for 6 preempted requests.
- Qwen3-8B: 32/32 requests completed; all 5 preempted requests recovered KV from LMCache.

Tested with:

```text
pytest -q tests/v1/test_mp_preemption_loading.py tests/v1/test_decode_save_and_preemption.py
21 passed
```

**If applicable**:

- [ ] this PR contains user facing changes - docs added
- [x] this PR contains unit tests

### CODE REVIEW

**Comment by malaiwah** [comment]
Thanks for closing the MP parity gap. I reviewed head `83325e32` against the surrounding MP tracker/load/store lifecycle and a TP4/DCP4 long-context deployment.

### Findings

The scheduler-side part looks directionally correct:

- rebuilding the tracker from `request.all_token_ids` allows a resumed request to include both prompt and generated tokens;
- retaining a fresh `PREFETCHING` tracker while an async lookup is polled avoids resubmission/reset;
- the full-hit `-1` adjustment is consistent with vLLM needing one token to recompute logits;
- the H20 preemption results are encouraging.

I do see one correctness concern that should be resolved or explicitly scoped before this is treated as a substrate for deliberate/cooperative preemption.

### Store durability / block-reuse concern on the default CUDA path

`build_connector_meta()` correctly raises `need_flush_before_forward` when requests are preempted/resumed, and `LMCacheMPWorkerAdapter.handle_preemptions()` calls `transfer_ctx.flush_inflight_stores()` before the next forward can overwrite freed paged-KV blocks.

However, CUDA `AUTO` selects `LMCacheDrivenTransferContext`, whose `flush_inflight_stores()` implementation is currently a no-op:

- [preemption signal and connector metadata](https://github.com/LMCache/LMCache/blob/83325e32c909c1afce04a2987f0856f9456e739d/lmcache/integration/vllm/lmcache_mp_connector.py)
- [transfer-context selection and LMCache-driven no-op flush](https://github.com/LMCache/LMCache/blob/83325e32c909c1afce04a2987f0856f9456e739d/lmcache/v1/multiprocess/transfer_context/worker_transfer.py)
- [worker-side preemption handling](https://github.com/LMCache/LMCache/blob/83325e32c909c1afce04a2987f0856f9456e739d/lmcache/integration/vllm/vllm_multi_process_adapter.py)

A `STORE` future can therefore still represent server-side reads from the registered GPU block IDs when vLLM preempts the owner, frees those IDs, and reuses them in the next forward. PR #4612 makes a later reload possible, but it does not by itself prove that the chunks being reloaded were durably captured before block reuse.

Could you establish one of these invariants for the LMCache-driven path?

1. `STORE` completion/ordering already guarantees that the server has finished reading every affected block before vLLM may reuse it; or
2. preemption waits for an all-rank store fence; or
3. the relevant blocks are pinned/deferred-free until all-rank completion.

This is analogous to the async external-KV corruption fixed in vLLM [#52360](https://github.com/vllm-project/vllm/issues/52360) / [#52372](https://github.com/vllm-project/vllm/pull/52372): store submission is not sufficient if the producer can reuse paged storage before the consumer has completed its read.

If the intended scope here is only **reload already-durable chunks**, I think that should be stated explicitly so this PR is not assumed to make arbitrary preemption storage-safe.

### Suggested coverage

The new tests validate lookup/tracker arithmetic, but they do not exercise allocation, retrieve metadata, worker transfer, completion, or block lifetime. I suggest adding or following up with:

- in-flight LMCache-driven store → preemption → immediate reuse of the same block IDs;
- delayed completion on one TP rank and an all-rank completion/failure case;
- repeated preemption of the same request;
- partial LMCache hit overlapping a nonzero local APC hit;
- preempted zero-hit/miss lifecycle;
- full-hit retrieve followed by last-token recomputation through the real metadata path.

Also, the Actions runs for this head currently show `action_required`, so CI has not independently exercised the patch yet.

I filed [vLLM #53277](https://github.com/vllm-project/vllm/issues/53277) to track cache-aware waiting QoS separately from the more demanding durable-checkpoint/cooperative-preemption phase. This PR is an important reload component of that second phase, provided the store-lifetime invariant above is made explicit and tested.

**Comment by Carlos779988** [comment]
Thanks for raising this. I ran a targeted H20 stress test on the ordering between LMCache-driven STORE operations and GPU block reuse. It confirmed that a GPU block can be released and reassigned while a STORE for that block is still pending. I also tested a provisional fence that waited on the STORE futures available when preemption was handled. That was not sufficient, because multiple STORE operations may share the same request ID, and another STORE can be registered after the fence has already selected which futures to wait on.

I will keep #4612 scoped to reloading chunks that are already durable in LMCache. I will update the PR description to state explicitly that this change does not make arbitrary or cooperative preemption storage-safe. The code and tests in this PR will remain limited to the current scheduler-side reload path.

I will follow up with a separate PR for STORE lifetime safety and broader worker/metadata lifecycle coverage, after aligning on the operation-level contract. That work will define operation-level completion, the block-release boundary, and all-rank completion/failure handling, and will cover the reuse and delayed-rank cases you listed. Does this split work for you?

**Comment by Carlos779988** [comment]
Following up on this PR. It is now scoped to reloading KV chunks that are already durable in LMCache; STORE lifetime safety is being handled separately. The current checks pass. Could one of the MP/vLLM maintainers take a look at the scoped change?
